syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.trippapp.Account";
option java_outer_classname = "AccountProto";
option objc_class_prefix = "ACP";

package Account;

service AccountService {
    rpc login (LoginReq) returns (LoginResp) {
    }
    rpc signup (SignupReq) returns (SignupResp) {
    }
    rpc get_user (UserReq) returns (UserResp) {
    }
    rpc get_account (AccountReq) returns (AccountResp) {
    }
    rpc get_followers (UserReq) returns (FollowersResp) {
    }
    rpc get_following (UserReq) returns (FollowingResp) {
    }
    rpc change_user (ChangeUserReq) returns (UserSummary) {
    }
    rpc change_profile (ChangeProfileReq) returns (UserSummary) {
    }
    rpc follow (FollowReq) returns (FollowResp) {
    }
    rpc unfollow (UnfollowReq) returns (UnfollowResp) {
    }
}


message Account {
    message User {
        int64 user_id = 1;
        string username = 3;
        string email = 6;
        bool can_login = 9;
    }
    message Profile {
        message Counts {
            int32 followers = 1;
            int32 following = 3;
            int32 posts = 6;
        }
        string pic_url = 1;
        string first_name = 3;
        string last_name = 6;
        bool verified = 9;
        string bio = 12;
        Counts counts = 15;
    }
    User user = 1;
    Profile profile = 3;
}
message ChangeAccount {
    message User {
        string username = 3;
        string email = 6;
    }
    message Password {
        string old_password = 1;
        string new_password = 3;
    }
    message Profile {
        string first_name = 1;
        string last_name = 3;
        string bio = 6;
    }
    message ProfilePic {
        string new_pic_url = 1;
    }
}
message UserSummary {
    int64 user_id = 1;
    string username = 3;
    string full_name = 6;
    string pic_url = 7;
}

/* Request Messages */
message LoginReq {
    string username_email = 1;
    string password = 6;
}
message SignupReq {
    string username = 1;
    string email = 3;
    string password = 6;
    string first_name = 9;
    string last_name = 12;
}
message UserReq {
    int64 user_id = 1;
    string username = 3;
    string email = 6;
}
message AccountReq {
    string user_id = 1;
    string self_user_id = 3;
}
message ChangeUserReq {
    string user_id = 1;
    ChangeAccount.User user = 3;
}
message ChangeProfileReq {
    string user_id = 1;
    ChangeAccount.Profile profile = 3;
}
message FollowReq {
    int64 self_user_id = 1;
    string self_username = 3;
    int64 user_id = 6;
    string username = 9;
}
message UnfollowReq {
    int64 self_user_id = 1;
    string self_username = 3;
    int64 user_id = 6;
    string username = 9;
}

/* Response Messages */
message RawMessage {
    bool succuss = 1;
    string message = 3;
}
message LoginResp {
    RawMessage result = 1;
    UserSummary user = 3;
    string token = 6;
}
message SignupResp {
    RawMessage result = 1;
    UserSummary user = 3;
    string token = 6;
}
message UserResp {
    RawMessage result = 1;
    UserSummary user = 3;
}
message AccountResp {
    RawMessage result = 1;
    Account account = 3;
}
message FollowersResp {
    RawMessage result = 1;
    repeated UserSummary followers = 3;
    int32 page_number = 6;
    int32 result_per_page = 9;
}
message FollowingResp {
    RawMessage result = 1;
    repeated UserSummary following = 3;
    int32 page_number = 6;
    int32 result_per_page = 9;
}
message FollowResp {
    RawMessage result = 1;
}
message UnfollowResp {
    RawMessage result = 1;
}